codice/ddf
# DDF-5303 Support query mappability in WFS 2.0 source

#### What does this PR do?
The WFS 2.0.0 source now supports the use of an optional MetacardMapper service to convert metacard attributes to WFS feature properties for queries to the WFS server.

This PR also fixes the WFS 2.0.0 connected source bundle so it is now usable again.

#### Who is reviewing it? 
@bennuttle 
@leo-sakh 
@gjvera 

#### Select relevant component teams: 
@codice/core-apis 
@codice/ogc 

#### Ask 2 committers to review/merge the PR and tag them here.
@kcwire 
@figliold 

#### How should this be tested?
1. Download [GeoServer](http://geoserver.org/).
2. Unzip it and run `java -jar start.jar`.
3. Follow the instructions [here](https://docs.geoserver.org/latest/en/user/gettingstarted/shapefile-quickstart/index.html) to set up GeoServer with some sample data.
a. Don't use `nyc_roads` as your data source (skip the `Data preparation` step).
b. In the `Create a store` step, use `<geoserver_install_location>/data_dir/data/nyc/poly_landmarks.shp` as the shapefile location. You can also name the data store something other than 'NYC Roads' but it doesn't really matter.
4. Install DDF.
5. In the Admin UI, go to the `Spatial` app and create a `Metacard to WFS Feature Map` with the following configuration:
a. `Feature Type`: {http://geoserver.org/nyc}poly_landmarks
b. `Metacard Title to WFS Feature Property Mapping`: LANAME
c. `Metacard Geography to WFS Feature Property Mapping`: the_geom
6. Create a WFS 2.0.0 Federated Source with the following configuration:
a. `WFS URL`: http://localhost:8080/geoserver/ows
b. `Forced Feature Type`: poly_landmarks
c. `SRS Name`: urn:ogc:def:crs:EPSG::4326
7. Open Intrigue and run a * query against the source. Verify the results show up on the map around New York City.
8. Run some queries against the `title` attribute, e.g. `CONTAINS *Park` or `= United Nations Headquarters`, and verify that the results look right.
a. This verifies that the metacard title mapping you created in step 5 is working.
9. Using the advanced query builder, add a new field using AND and set the attribute to either `anyGeo` or `location`. Zoom in on New York City and draw a bounding box that contains a subset of the results currently on the map. Run the query and verify the results are inside the box.
a. This verifies that the metacard location mapping you created in step 6 is working.
10. Delete the WFS 2.0.0 Federated Source and repeat steps 6-9 using a WFS 2.0.0 Connected Source.
a. In Intrigue you'll have to query the `ddf.distribution` source since your connected source will be hidden.

#### Any background context you want to provide?
https://github.com/codice/ddf/pull/5091 removed the dynamically-generated metacard attributes of the form `ext.FeatureType.FeatureProperty` so those can no longer be used to query WFS sources. Users will have to configure metacard attribute to WFS feature property mappings via a `MetacardMapper` and query their WFS source using metacard attributes.

#### What are the relevant tickets?
Fixes: #5303

#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
